src/utils/wakers/shared_arc.rs

```diff
@@ -0,0 +1,92 @@
+//    │    / ┌─────────────┬──────┼───────┬──────┼───────┬──────┼───────┬─────┐ \
+//    ▼   /  │             │      │       │      │       │      │       │     │  \
+//   Arc <   │  Readiness  │ wake_data[0] │ wake_data[1] │ wake_data[2] │ ... │   >
+//    ▲   \  │             │              │              │              │     │  /
+//    │    \ └─────────────┴──────▲───────┴──────▲───────┴──────▲───────┴─────┘ /
+//    │                           │              │              │
+//    └─┐         ┌───────────────┘              │              │
+//      │         │                              │              │
+//      │         │           ┌──────────────────┘              │
+//      │         │           │                                 │
+//      │         │           │           ┌─────────────────────┘
+//      │         │           │           │
+//      │         │           │           │
+// ┌────┼────┬────┼──────┬────┼──────┬────┼──────┬─────┐
+// │    │    │    │      │    │      │    │      │     │
+// │  Inner  │ wakers[0] │ wakers[1] │ wakers[2] │ ... │
+// │         │           │           │           │     │
+// └─────────┴───────────┴───────────┴───────────┴─────┘
+
+// TODO: Right now each waker gets its own wake_data slot.
+// We can save space by making size_of::<usize>() wakers share the same slot.
+// With such change, in 64-bit system, the wake_data array/vec would only need ⌈N/8⌉ slots instead of N.
+
+pub(super) trait WakeDataContainer {
+    /// Get the reference of the wake_data slice. This is used to compute the index.
+    fn get_wake_data_slice(&self) -> &[*const Self];
+    /// Called when the subfuture at the specified index should be polled.
+    fn wake_index(&self, index: usize);
+}
+pub(super) unsafe fn waker_for_wake_data_slot<A: WakeDataContainer>(
+    pointer: *const *const A,
+) -> Waker {
+    unsafe fn clone_waker<A: WakeDataContainer>(pointer: *const ()) -> RawWaker {
+        let pointer = pointer as *const *const A;
+        let raw = *pointer; // This is the raw pointer of Arc<Inner>.
+
+        // We're creating a new Waker, so we need to increment the count.
+        Arc::increment_strong_count(raw);
+
+        RawWaker::new(pointer as *const (), create_vtable::<A>())
+    }
+
+    // Convert a pointer to a wake_data slot to the Arc<Inner>.
+    unsafe fn to_arc<A: WakeDataContainer>(pointer: *const *const A) -> Arc<A> {
+        let raw = *pointer;
+        Arc::from_raw(raw)
+    }
+    unsafe fn wake<A: WakeDataContainer, const BY_REF: bool>(pointer: *const ()) {
+        let pointer = pointer as *const *const A;
+        let arc = to_arc::<A>(pointer);
+        // Calculate the index
+        let index = ((pointer as usize) // This is the slot our pointer points to.
+            - (arc.get_wake_data_slice() as *const [*const A] as *const () as usize)) // This is the starting address of wake_data.
+            / std::mem::size_of::<*const A>();
```

Comment:
Rather than adding comments, can I ask you to assign these to descriptive variables instead? e.g. 
```suggestion
        let pointer_slot = (pointer as usize);
        let wake_data_slice = arc.get_wake_data_slice();
        let start_addr = (wake_data_slice as *const [*const A] as *const () as usize);
        let index = (pointer_slot - start_addr) / std::mem::size_of::<*const A>();
```

I'm not super sure why `std::mem::size_of::<*const A>()` exists, but maybe that too can be assigned to a variable?

Reply:
Also, now that I've written this out like this - we should probably not perform pointer-to-usize casts. The stdlib has been rewritten to move away from that, meaning we can't PR this code the stdlib right now. Instead we should probably use the [strict provenance polyfill](https://github.com/Gankra/sptr) to perform pointer arithmatic.

Reply:
Added the polyfill. Luckily we weren't doing any usize->pointer cast so there's not much to worry about regarding provenance.